atoma-proxy/src/server/mod.rs

```diff
@@ -0,0 +1,220 @@
+            error!("Failed to get GetStacksForModel result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?;
+
+    let selected_node_id = if stacks.is_empty() {
+        let (result_sender, result_receiver) = oneshot::channel();
+        state
+            .state_manager_sender
+            .send(AtomaAtomaStateManagerEvent::GetTasksForModel {
+                model: model.clone(),
+                result_sender,
+            })
+            .map_err(|err| {
+                error!("Failed to send GetTasksForModel event: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?;
+        let tasks = result_receiver
+            .await
+            .map_err(|err| {
+                error!("Failed to receive GetTasksForModel result: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?
+            .map_err(|err| {
+                error!("Failed to get GetTasksForModel result: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?;
+        if tasks.is_empty() {
+            error!("No tasks found for model {}", model);
+            return Err(StatusCode::NOT_FOUND);
+        }
+        state
+            .sui
+            .write()
+            .await
+            .acquire_new_stack_entry(tasks[0].task_small_id as u64, 1000, 100)
+            .await
+            .map_err(|err| {
+                error!("Failed to acquire new stack entry: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?
+    } else {
+        stacks[0].selected_node_id
+    };
+
+    let (result_sender, result_receiver) = oneshot::channel();
+    state
+        .state_manager_sender
+        .send(AtomaAtomaStateManagerEvent::GetNodePublicAddress {
+            node_small_id: selected_node_id,
+            result_sender,
+        })
+        .map_err(|err| {
+            error!("Failed to send GetNodePublicAddress event: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?;
+
+    let node_address = result_receiver
+        .await
+        .map_err(|err| {
+            error!("Failed to receive GetNodePublicAddress result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .map_err(|err| {
+            error!("Failed to get GetNodePublicAddress result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .ok_or_else(|| {
+            error!("No node address found for node {}", selected_node_id);
+            StatusCode::NOT_FOUND
+        })?;
+
+    state
+        .sui
+        .write()
+        .await
+        .send_openai_api_request(payload, node_address)
+        .await
+        .map_err(|err| {
+            error!("Failed to send OpenAI API request: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })
+        .map(|response| Json(response))
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct NodePublicIpAssignment {
```

Comment:
I suggest the following:

1. Add docs
2. Change the above logic to different files, to make it easier to search in the codebase, later on.